[flang][OpenACC][OpenMP] Separate implementations of ATOMIC constructs

kparzysz opened the pull request:
The OpenMP implementation of the ATOMIC construct will change in the near future to accommodate OpenMP 6.0. This patch separates the shared implementations to avoid interfering with OpenACC.

kparzysz committed:
[flang][OpenMP] Mark atomic clauses as unique

The current implementation of the ATOMIC construct handles these clauses
individually, and this change does not have an observable effect. At the
same time these clauses are unique as per the OpenMP spec, and this patch
reflects that in the OMP.td file.

kparzysz committed:
[flang][OpenACC][OpenMP] Separate implementations of ATOMIC constructs

The OpenMP implementation of the ATOMIC construct will change in the
near future to accommodate OpenMP 6.0. This patch separates the shared
implementations to avoid interfering with OpenACC.

kparzysz commented:
Previous PR: https://github.com/llvm/llvm-project/pull/137460
Next PR: https://github.com/llvm/llvm-project/pull/137521

tblah reviewed (approved):
LGTM. If I didn't miss anything, this patch copies the old header code verbatim into both implementation files?

kparzysz commented:
Yes, it started as a copy, then I removed templates, and replaced type parameters with concrete types, and removed the "OmpAcc" from function names.

clementval reviewed (approved):
LGTM

razvanlupusoru reviewed (approved):
Would you mind clarifying which atomic features from OpenMP 6.0 cause the difference between OpenMP and OpenACC atomics to widen that they need separated out?

kparzysz commented:
> Would you mind clarifying which atomic features from OpenMP 6.0 cause the difference between OpenMP and OpenACC atomics to widen that they need separated out?

Short answer: conditional-update and change in the implementation of the existing OpenMP code.

The conditional update (ATOMIC COMPARE) actually exists in OpenMP prior to 6.0, but hasn't been implemented in flang (beyond parsing).  Without being aware of that yet, I tried to implement conditional update-capture (with post-6.0 clarifications), and then I realized that it's not going to be a simple addition of extra code.

I'm changing the OpenMP implementation in some fundamental ways.  In particular, I'm putting the majority of work (to analyze and validate) the AST nodes into the semantic analysis, and within it, the expressions and assignments will be universally represented by `evaluate::Expr` and `evaluate::Assignment`.  The existing shared code used AST nodes as the sources for MLIR-gen, the new code will use "evaluate".  There is no `OmpAtomicClause` anymore, there is a single `OpenMPAtomicConstruct` for everything, and so on.  This essentially replaced the OpenMP versions of the shared functions, so I decided to remove the rest from the header file and put it in OpenACC.cpp (which was the only user of that code).

This patch simply extracts that step into an individual PR to reduce the size of the upcoming OpenMP PR.

After everything is done, it may be possible for OpenACC to reuse the code, but in the meantime I didn't want to modify OpenACC as a part of this work.

Going back to the conditional-update: I'm planning to change the OpenMP atomic MLIR ops to avoid having to add an extra one for conditional-update-capture, so that may affect sharing in some way.

kparzysz commented:
I updated the commit message to be more specific about the goals.
